File: c/test_nimpy.nim
# test_nimpy.nim - file name should match the module name you're going to import from python
import nimpy
import json

proc greet(name: string): string {.exportpy.} =
  return "Hello, " & name & "!"

#[
    nimble install nimpy
    nim c --app:lib --out:test_nimpy.so --threads:on test_nimpy
]#


proc check_py_call(): string {.exportpy.} =
  let os = pyImport("os")
  echo "Current dir is: ", os.getcwd().to(string)

# sum(range(1, 5))
  let py = pyBuiltinsModule()
  let s = py.sum(py.range(0, 5)).to(int)
  assert s == 10

  return $(s)


proc tupleDiff(a, b: tuple[x, y: int]): tuple[x, y: int] {.exportpy: "tuple_diff".} =
  result = (a.x - b.x, a.y - b.y)


proc nimOutConfig(): JsonNode {.exportpy: "nim_json".} =
  result = %* { "init" : 1.0,
                "final" : 5,
                "nested" : [1, 2, 3.5, {"InArray" : 5}],
                "dictLike" : { "Nested" : "Value" }
              }

iterator testIterator(s: string): int {.exportpy: "iter_from_nim".} =
  for i in 0 ..< s.len:
    yield i


type
  Obj = ref object of PyNimObjectExperimental
    a, b: int
    c: string

  MyRefObj = ref Obj

proc validateMyObj(o: MyRefObj): bool {.exportpy: "validate_obj".} =
  o.a == 5 and o.b == 3 and o.c == "hello"

File: c/test_nimpy.py
# test.py
import test_nimpy as mymodule
assert mymodule.greet("world") == "Hello, world!" # type: ignore
assert mymodule.greet(name="world") == "Hello, world!" # type: ignore

print("Passed all tests here...")
print(f"{mymodule.greet(name='world')=}") # type: ignore


print(f"{mymodule.check_py_call()}") # type: ignore


t1 = (1,2)
t2 = (3,4)

print(f"{mymodule.tuple_diff(t1, t2)}") # type: ignore

import timeit

a = timeit.timeit("""
mymodule.tuple_diff(t1, t2)
""", number=10000, setup='import test_nimpy as mymodule; t1 = (1,2); t2 = (3,4);')

b = timeit.timeit("""
(t1[i] - t2[i] for i in range(2))
""", number=10000)

print(f"{a=}, {b=}")

print(mymodule.nim_json()) # type: ignore


from dataclasses import dataclass

@dataclass
class Obj:
    a: int
    b: int
    c: str


a_test = Obj(a=5, b=3, c="hello")

print(a_test)
print(mymodule.validate_obj(a_test)) # type: ignore
## ODDLY this is false, because we have to reference
#  and create the object internal to Nim for it to correctly validate?

# Need to import as a py object??